File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/ActionDatabase.java
package versioned.host.exp.exponent.modules.api.notifications;

import android.arch.persistence.db.SupportSQLiteOpenHelper;
import android.arch.persistence.room.Database;
import android.arch.persistence.room.DatabaseConfiguration;
import android.arch.persistence.room.InvalidationTracker;
import android.arch.persistence.room.RoomDatabase;
import android.support.annotation.NonNull;

@Database(entities = {ActionObject.class}, version = 1, exportSchema = false)
public abstract class ActionDatabase extends RoomDatabase {
  public abstract ActionObjectDao mActionObjectDao();
}

File: apps/expo-go/android/expoview/src/main/java/host/exp/exponent/notifications/ActionDatabase.kt
package host.exp.exponent.notifications

import android.content.Context
import androidx.room.Dao
import androidx.room.Database
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Room
import androidx.room.RoomDatabase

@Dao
interface ActionDao {
  @Insert(onConflict = OnConflictStrategy.REPLACE)
  fun insert(action: ActionObject)

  @Query("SELECT * FROM ActionObject WHERE categoryId = :categoryId ORDER BY position ASC")
  fun findByCategory(categoryId: String?): List<ActionObject>

  @Query("DELETE FROM ActionObject WHERE categoryId = :categoryId")
  fun deleteByCategory(categoryId: String?)
}

@Database(entities = [ActionObject::class], version = ActionDatabase.VERSION, exportSchema = false)
abstract class ActionRoomDatabase : RoomDatabase() {
  abstract fun actionDao(): ActionDao
}

object ActionDatabase {
  const val NAME = "ExpoNotificationActions"
  const val VERSION = 2

  lateinit var dao: ActionDao
    private set

  @JvmStatic fun init(context: Context) {
    dao = Room.databaseBuilder(context, ActionRoomDatabase::class.java, "$NAME.db")
      .fallbackToDestructiveMigration() 
      .allowMainThreadQueries()
      .build()
      .actionDao()
  }
}
